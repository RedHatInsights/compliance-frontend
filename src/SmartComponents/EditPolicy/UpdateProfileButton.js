import React from 'react';
import { graphql } from '@apollo/react-hoc';
import { Button } from '@patternfly/react-core';
import propTypes from 'prop-types';
import { CREATE_BUSINESS_OBJECTIVE, UPDATE_PROFILE } from '../../Utilities/graphql/mutations';

export class UpdateProfileButton extends React.Component {
    handleBusinessObjective = () => {
        const { businessObjective, editPolicyBusinessObjective, mutate } = this.props;

        if (editPolicyBusinessObjective === undefined) {
            return Promise.resolve(businessObjective ? businessObjective.id : null);
        }

        if (editPolicyBusinessObjective && !editPolicyBusinessObjective.create && businessObjective
            && (editPolicyBusinessObjective.value !== businessObjective.id)) {
            return Promise.resolve(editPolicyBusinessObjective.value);
        }

        if (editPolicyBusinessObjective && !editPolicyBusinessObjective.create
            && businessObjective === null) {
            return Promise.resolve(editPolicyBusinessObjective.value);
        }

        if (editPolicyBusinessObjective === null) {
            return Promise.resolve(null);
        }

        if (editPolicyBusinessObjective.create) {
            return mutate({
                mutation: CREATE_BUSINESS_OBJECTIVE,
                variables: { input: { title: editPolicyBusinessObjective.label } }
            }).then((result) => {
                return result.data.createBusinessObjective.businessObjective.id;
            });
        }
    }

    onClick = () => {
        const { mutate, policyId, threshold, onClick } = this.props;

        return this.handleBusinessObjective().then((businessObjectiveId) => {
            let input = {
                id: policyId,
                complianceThreshold: parseFloat(threshold)
            };

            if (businessObjectiveId) {
                input.businessObjectiveId = businessObjectiveId;
            }

            return mutate({
                mutation: UPDATE_PROFILE,
                variables: { input }
            });
        }).then(() => {
            onClick();
        });
    }

    render() {
        return (<Button type='submit' variant='primary'
            onClick={this.onClick}>Save</Button>
        );
    }
}

UpdateProfileButton.propTypes = {
    policyId: propTypes.string.isRequired,
    businessObjective: propTypes.object,
    editPolicyBusinessObjective: propTypes.object,
    mutate: propTypes.func.isRequired,
    threshold: propTypes.number,
    onClick: propTypes.func.isRequired
};

export default graphql(UPDATE_PROFILE)(UpdateProfileButton);
