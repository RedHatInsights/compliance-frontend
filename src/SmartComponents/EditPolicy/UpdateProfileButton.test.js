import { UpdateProfileButton } from './UpdateProfileButton';

describe('UpdateProfileButton', () => {
    const defaultProps = {
        businessObjective: {
            title: 'Test Objective',
            id: '1'
        },
        policyId: '12345',
        editPolicyBusinessObjective: {
            title: 'Edited Test Objective',
            value: '2'
        },
        mutate: jest.fn(),
        onClick: jest.fn()
    };

    it('expect to render without error', () => {
        const wrapper = shallow(
            <UpdateProfileButton { ...defaultProps } />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    describe('.handleBusinessObjective', () => {
        it('returns the businessObjective.id if editPolicyBusinessObjective is undefined', () => {
            const wrapper = shallow(
                <UpdateProfileButton { ...defaultProps }
                    editPolicyBusinessObjective={ undefined } />
            );
            const instance = wrapper.instance();

            instance.handleBusinessObjective().then((resultId) => {
                expect(resultId).toBe(defaultProps.businessObjective.id);
            });
        });

        it('returns null if editPolicyBusinessObjective is undefined and no businessObjective', () => {
            const wrapper = shallow(
                <UpdateProfileButton { ...defaultProps }
                    businessObjective={ undefined }
                    editPolicyBusinessObjective={ undefined } />
            );
            const instance = wrapper.instance();

            instance.handleBusinessObjective().then((resultId) => {
                expect(resultId).toBe(null);
            });
        });

        it('returns the editPolicyBusinessObjective.value if it is defined and businessObjective is null', () => {
            const wrapper = shallow(
                <UpdateProfileButton { ...defaultProps }
                    businessObjective={ null } />
            );
            const instance = wrapper.instance();

            instance.handleBusinessObjective().then((resultId) => {
                expect(resultId).toBe(defaultProps.editPolicyBusinessObjective.value);
            });
        });

        it('returns the editPolicyBusinessObjective.value if it and businessObjective is defined', () => {
            const wrapper = shallow(
                <UpdateProfileButton { ...defaultProps } />
            );
            const instance = wrapper.instance();

            instance.handleBusinessObjective().then((resultId) => {
                expect(resultId).toBe(defaultProps.editPolicyBusinessObjective.value);
            });
        });

        it('returns null when editPolicyBusinessObjective is null', () => {
            const wrapper = shallow(
                <UpdateProfileButton { ...defaultProps }
                    editPolicyBusinessObjective={ null } />
            );
            const instance = wrapper.instance();

            instance.handleBusinessObjective().then((resultId) => {
                expect(resultId).toBe(null);
            });
        });

        it('returns a new ID if editPolicyBusinessObjective has a create property and mutates', async () => {
            const title = 'New Objective';
            const mutate = jest.fn((query) => {
                let id;

                if (query.variables.input.title === title)  {
                    id = '3';
                } else {
                    id = null;
                }

                return Promise.resolve({
                    data: {
                        createBusinessObjective: {
                            businessObjective: {
                                id
                            }
                        }
                    }
                });
            });
            const wrapper = shallow(
                <UpdateProfileButton { ...defaultProps }
                    mutate={ mutate }
                    businessObjective={ undefined }
                    editPolicyBusinessObjective={ {
                        create: true,
                        label: title
                    } } />
            );
            const instance = wrapper.instance();

            await instance.handleBusinessObjective().then((resultId) => {
                expect(resultId).toBe('3');
            });
            expect(mutate).toHaveBeenCalled();
        });
    });

    describe('.onClick', () => {
        it('calls mutate and onClick', async () => {
            const mutate = jest.fn();
            const hBO = jest.fn(() => Promise.resolve(1));
            const wrapper = shallow(
                <UpdateProfileButton { ...defaultProps } mutate={ mutate } />
            );
            wrapper.instance().handleBusinessObjective = hBO;
            wrapper.update();

            await wrapper.instance().onClick();

            expect(mutate).toHaveBeenCalled();
            expect(hBO).toHaveBeenCalled();
        });
    });
});
