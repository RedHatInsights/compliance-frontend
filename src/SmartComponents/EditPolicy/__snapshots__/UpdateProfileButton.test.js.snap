// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateProfileButton expect to render without error 1`] = `
<Component
  onClick={[Function]}
  type="submit"
  variant="primary"
>
  Save
</Component>
`;
